File: v4/rust/iprange-livedb/benches/update_ipsets/measure.rs
use std::fs;
use std::io::Write;
use std::path::Path;
use std::time::{Duration, Instant};

use crate::allocation::{self, AllocationStats};

#[derive(Clone, Copy, Debug, Default)]
pub(crate) struct FileSize {
    pub(crate) logical: u64,
    pub(crate) physical: Option<u64>,
}

#[derive(Clone, Debug)]
pub(crate) struct Measurement {
    pub(crate) elapsed: Duration,
    pub(crate) allocations: AllocationStats,
    pub(crate) rss_before_kib: Option<u64>,
    pub(crate) rss_after_kib: Option<u64>,
    pub(crate) rss_peak_kib: Option<u64>,
    pub(crate) fds_before: Option<u64>,
    pub(crate) fds_after: Option<u64>,
}

pub(crate) fn operation<T>(callback: impl FnOnce() -> T) -> (T, Measurement) {
    let rss_before_kib = current_rss_kib();
    let fds_before = open_file_descriptors();
    profiler_command(b"enable\n");
    let started = Instant::now();
    let (result, allocations) = allocation::measure(callback);
    let elapsed = started.elapsed();
    profiler_command(b"disable\n");
    let fds_after = open_file_descriptors();
    let rss_after_kib = current_rss_kib();
    (
        result,
        Measurement {
            elapsed,
            allocations,
            rss_before_kib,
            rss_after_kib,
            rss_peak_kib: peak_rss_kib(),
            fds_before,
            fds_after,
        },
    )
}

fn profiler_command(command: &[u8]) {
    let Some(path) = std::env::var_os("IPRANGE_PERF_CONTROL") else {
        return;
    };
    let mut control = fs::OpenOptions::new()
        .write(true)
        .open(&path)
        .unwrap_or_else(|error| panic!("open perf control {path:?}: {error}"));
    control
        .write_all(command)
        .unwrap_or_else(|error| panic!("write perf control {path:?}: {error}"));
}

pub(crate) fn file_size(path: &Path) -> std::io::Result<FileSize> {
    let metadata = fs::metadata(path)?;
    Ok(FileSize {
        logical: metadata.len(),
        physical: physical_bytes(&metadata),
    })
}

#[cfg(unix)]
fn physical_bytes(metadata: &fs::Metadata) -> Option<u64> {
    use std::os::unix::fs::MetadataExt;

    Some(metadata.blocks().saturating_mul(512))
}

#[cfg(not(unix))]
fn physical_bytes(_: &fs::Metadata) -> Option<u64> {
    None
}

#[cfg(target_os = "linux")]
fn current_rss_kib() -> Option<u64> {
    status_value_kib("VmRSS:")
}

#[cfg(not(target_os = "linux"))]
fn current_rss_kib() -> Option<u64> {
    None
}

#[cfg(target_os = "linux")]
fn peak_rss_kib() -> Option<u64> {
    status_value_kib("VmHWM:")
}

#[cfg(all(unix, not(target_os = "linux")))]
fn peak_rss_kib() -> Option<u64> {
    let mut usage = unsafe { std::mem::zeroed::<libc::rusage>() };
    if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut usage) } != 0 {
        return None;
    }
    let value = u64::try_from(usage.ru_maxrss).ok()?;
    #[cfg(target_os = "macos")]
    {
        Some(value / 1024)
    }
    #[cfg(not(target_os = "macos"))]
    {
        Some(value)
    }
}

#[cfg(not(unix))]
fn peak_rss_kib() -> Option<u64> {
    None
}

#[cfg(target_os = "linux")]
fn status_value_kib(label: &str) -> Option<u64> {
    let status = fs::read_to_string("/proc/self/status").ok()?;
    status.lines().find_map(|line| {
        line.strip_prefix(label)?
            .split_ascii_whitespace()
            .next()?
            .parse()
            .ok()
    })
}

#[cfg(target_os = "linux")]
fn open_file_descriptors() -> Option<u64> {
    Some(fs::read_dir("/proc/self/fd").ok()?.count() as u64)
}

#[cfg(not(target_os = "linux"))]
fn open_file_descriptors() -> Option<u64> {
    None
}

File: v4/go/cmd/iprange-v4-bench/measure.go
// Measurement protocol (Rust benches/update_ipsets/measure.rs +
// timing.rs + allocation.rs): elapsed time, allocation counters through
// runtime MemStats Mallocs/TotalAlloc deltas, RSS and peak RSS from
// /proc/self/status on Linux, open fd count from /proc/self/fd, and
// logical plus physical file size. The optional
// IPRANGE_PERF_CONTROL/IPRANGE_PERF_ACK protocol matches Rust.
package main

import (
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type allocationStats struct {
	calls uint64
	bytes uint64
}

type fileSize struct {
	logical  uint64
	physical *uint64
}

type measurement struct {
	elapsed      time.Duration
	allocations  allocationStats
	rssBeforeKib *uint64
	rssAfterKib  *uint64
	rssPeakKib   *uint64
	fdsBefore    *uint64
	fdsAfter     *uint64
}

func memAllocationStats() allocationStats {
	var stats runtime.MemStats
	runtime.ReadMemStats(&stats)
	return allocationStats{calls: stats.Mallocs, bytes: stats.TotalAlloc}
}

func operation(callback func() error) (error, measurement) {
	perfCommand([]byte("enable\n"))
	rssBefore := currentRSSKib()
	fdsBefore := openFileDescriptors()
	// One GC before the baseline keeps the allocation delta attributable
	// to the measured operation.
	runtime.GC()
	started := time.Now()
	before := memAllocationStats()
	err := callback()
	after := memAllocationStats()
	elapsed := time.Since(started)
	fdsAfter := openFileDescriptors()
	rssAfter := currentRSSKib()
	perfCommand([]byte("disable\n"))
	return err, measurement{
		elapsed: elapsed,
		allocations: allocationStats{
			calls: after.calls - before.calls,
			bytes: after.bytes - before.bytes,
		},
		rssBeforeKib: rssBefore,
		rssAfterKib:  rssAfter,
		rssPeakKib:   peakRSSKib(),
		fdsBefore:    fdsBefore,
		fdsAfter:     fdsAfter,
	}
}

func perfCommand(command []byte) {
	path := os.Getenv("IPRANGE_PERF_CONTROL")
	if path == "" {
		return
	}
	control, err := os.OpenFile(path, os.O_WRONLY, 0)
	if err != nil {
		panic("open perf control " + path + ": " + err.Error())
	}
	if _, err := control.Write(command); err != nil {
		panic("write perf control " + path + ": " + err.Error())
	}
	_ = control.Close()
	ackPath := os.Getenv("IPRANGE_PERF_ACK")
	if ackPath == "" {
		return
	}
	ack, err := os.Open(ackPath)
	if err != nil {
		panic("open perf acknowledgement " + ackPath + ": " + err.Error())
	}
	defer ack.Close()
	response := make([]byte, 5)
	if _, err := ack.Read(response); err != nil {
		panic("read perf acknowledgement " + ackPath + ": " + err.Error())
	}
	if string(response) != "ack\n\x00" {
		panic("unexpected perf acknowledgement")
	}
}

func currentRSSKib() *uint64 { return statusValueKib("VmRSS:") }

func peakRSSKib() *uint64 { return statusValueKib("VmHWM:") }

func statusValueKib(label string) *uint64 {
	status, err := os.ReadFile("/proc/self/status")
	if err != nil {
		return nil
	}
	for _, line := range strings.Split(string(status), "\n") {
		if !strings.HasPrefix(line, label) {
			continue
		}
		fields := strings.Fields(strings.TrimPrefix(line, label))
		if len(fields) == 0 {
			return nil
		}
		value, err := strconv.ParseUint(fields[0], 10, 64)
		if err != nil {
			return nil
		}
		return &value
	}
	return nil
}

func openFileDescriptors() *uint64 {
	entries, err := os.ReadDir("/proc/self/fd")
	if err != nil {
		return nil
	}
	value := uint64(len(entries))
	return &value
}

// fileSizeOf mirrors Rust measure::file_size: logical length plus the
// physical block count (st_blocks x 512) where the platform reports it
// (unix Stat_t carries Blocks; other platforms report the logical size
// only, exactly like the Rust reported-size rules).
func fileSizeOf(path string) (fileSize, error) {
	info, err := os.Stat(path)
	if err != nil {
		return fileSize{}, err
	}
	size := fileSize{logical: uint64(info.Size())}
	if blocks := statBlocks(info); blocks > 0 {
		physical := uint64(blocks) * 512
		size.physical = &physical
	}
	return size, nil
}
